### PR TITLE
Laboratory work #1, Alina Tryasoumova - 23FPL1

### PR DESCRIPTION
**Pull request by AlinaTR52**

### CODE REVIEW

**Comment by EIzharikov** [comment]
Nothing to review, I wish you good luck and advise you to hurry up!

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -7,48 +7,28 @@
 
 
 def tokenize(text: str) -> list[str] | None:
-    """
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:10]: dont delete docstrings. it is forbidden

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -7,48 +7,28 @@
...
-        text (str): A text
-
-    Returns:
-        dict[str, str | dict[str, float]] | None: A dictionary with two keys – name, freq
-
-    In case of corrupt input arguments, None is returned
-    """
+    profile = {}
+    profile['name'] = language
+    frequency = calculate_frequencies(tokenize(text))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:29]: what if this func will return None?

File: seminars/practice_1_run_debug.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
 """
```
Comment (EIzharikov) [inline seminars/practice_1_run_debug.py:2]: revert changes in seminar files

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -7,48 +7,28 @@
...
-
+    letters = []
+    for letter in text:
+        if letter.isalpha():
+            letters.append(letter.lower())
+    return letters
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
-    """
-    Calculate frequencies of given tokens.
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:27]: .

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -7,48 +7,28 @@
...
+    return vocabulary
 
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
-    """
-    Create a language profile.
-
-    Args:
-        language (str): A language
-        text (str): A text
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:45]: .

File: lab_1_classify_by_unigrams/settings.json [hunk 6]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 4
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: I believe that you can better, you can do it!)

**Review by EIzharikov — COMMENTED** [review]
Nice! Make your CI green and continue to work. You can achieve six at least!

File: seminars/practice_1_run_debug.py (hunk 3 above)
Comment (AlinaTR52) [reply seminars/practice_1_run_debug.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:10]: where is the answer?

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -5,51 +5,78 @@
...
     """
-    Split a text into tokens.
+       Split a text into tokens.
 
-    Convert the tokens into lowercase, remove punctuation, digits and other symbols
+       Convert the tokens into lowercase, remove punctuation, digits and other symbols
 
-    Args:
-        text (str): A text
+       Args:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:14]: remove useless indents

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -5,51 +5,78 @@
...
-    Returns:
-        list[str] | None: A list of lower-cased tokens without punctuation
+       Returns:
+           list[str] | None: A list of lower-cased tokens without punctuation
 
-    In case of corrupt input arguments, None is returned
+       In case of corrupt input arguments, None is returned
     """
-
+    if (type(text)) != str:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:22]: use isinstance instead of type()

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -5,51 +5,78 @@
...
+        return None
+    letters = []
+    for letter in text:
+        if letter.isalpha():
+            letters.append(letter.lower())
+    if len(letters) == 0:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:28]: you can replace this by 
`if not letters:
    return None`

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -5,51 +5,78 @@
...
+    return letters
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     """
-    Calculate frequencies of given tokens.
+        Calculate frequencies of given tokens.
 
-    Args:
-        tokens (list[str] | None): A list of tokens
+        Args:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:36]: .

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -5,51 +5,78 @@
...
 
-    Returns:
-        dict[str, float] | None: A dictionary with frequencies
+        Returns:
+            dict[str, float] | None: A dictionary with frequencies
 
-    In case of corrupt input arguments, None is returned
+        In case of corrupt input arguments, None is returned
     """
+    if (type(tokens)) != list:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:44]: isinstance

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -5,51 +5,78 @@
...
+        return None
+    else:
+        for i in tokens:
+            if (type(i)) != str:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:48]: also isinstance

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:76]: where is the answer?

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -64,6 +91,15 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     In case of corrupt input arguments, None is returned
     """
+    if (type(predicted)) != list or (type(actual)) != list or len(predicted) != len(actual):
+        return None
+    else:
+        n = len(predicted)
+    summa = 0
+    for i in range(n):
+        summa += (actual[i]-predicted[i])**2
+    mse = summa/n
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:101]: return as is

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -84,6 +120,27 @@ def compare_profiles(
...
+    actual = []
+    predicted = []
+    for letter in vocabulary:
+        actual.append(unknown_profile['freq'].get(letter, 0))
+    for letter in vocabulary:
+        predicted.append(profile_to_compare['freq'].get(letter,0))
+
+    mse = calculate_mse(predicted=predicted, actual=actual)
+
+    return round(mse, 3)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:143]: why 3?

File: lab_1_classify_by_unigrams/settings.json [hunk 15]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 8
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: Nice aim!

File: lab_1_classify_by_unigrams/start.py [hunk 16]
```diff
@@ -2,21 +2,29 @@
...
     with open("assets/texts/de.txt", "r", encoding="utf-8") as file_to_read_de:
         de_text = file_to_read_de.read()
+        de_profile = create_language_profile(language='de', text=de_text)
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+        unknown_profile = create_language_profile(language='unk', text=unknown_text)
+
+    result = detect_language(unknown_profile=unknown_profile, profile_1=en_profile, profile_2=de_profile)
+    print(result)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:22]: is it working?

File: lab_1_classify_by_unigrams/start.py [hunk 17]
```diff
@@ -2,21 +2,29 @@
...
     assert result, "Detection result is None"
 
 
+
+
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:27]: remove whitespaces

**Review by EIzharikov — COMMENTED** [review]
Fix my comments, make CI green and be ready for the next review! Good luck!

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py:22]: For all functions?

File: lab_1_classify_by_unigrams/start.py (hunk 16 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/start.py:22]: Yes, it is

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: I thought we should round mse like in the example. Is it unessesary?

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:10]: ?...

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py]: You should know :)

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py:10]: fixed, sorry(

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -64,6 +87,13 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(predicted, list) or not isinstance(actual, list) or len(predicted) != len(actual):
+        return None
+    n = len(predicted)
+    summa = 0
```
[the remark this answers is not in the document]
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: Yes, have changed it))

File: lab_1_classify_by_unigrams/main.py [hunk 19]
```diff
@@ -84,6 +114,23 @@ def compare_profiles(
...
+    if not isinstance(unknown_profile, dict) or not isinstance(profile_to_compare, dict):
+        return None
+    if "name" not in unknown_profile or "name" not in profile_to_compare:
+        return None
+    if not isinstance(unknown_profile["freq"], dict) or not isinstance(profile_to_compare["freq"], dict):
+        return None
+    vocabulary = set(unknown_profile['freq'].keys()) | set(profile_to_compare['freq'].keys())
+    actual = []
+    predicted = []
+    for letter in vocabulary:
```
[the remark this answers is not in the document]
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py:129]: Yes, fixed

File: lab_1_classify_by_unigrams/main.py [hunk 20]
```diff
@@ -84,6 +114,23 @@ def compare_profiles(
...
+        actual.append(unknown_profile['freq'].get(letter, 0))
+    for letter in vocabulary:
+        predicted.append(profile_to_compare['freq'].get(letter,0))
+
+    mse = calculate_mse(predicted=predicted, actual=actual)
```
[the remark this answers is not in the document]
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/main.py]: fixed

File: lab_1_classify_by_unigrams/start.py [hunk 21]
```diff
@@ -2,19 +2,24 @@
...
 # pylint:disable=too-many-locals, unused-argument, unused-variable
-
-
+from lab_1_classify_by_unigrams.main import create_language_profile, detect_language
 def main() -> None:
     """
     Launches an implementation
     """
     with open("assets/texts/en.txt", "r", encoding="utf-8") as file_to_read_en:
+        #tokenized_text = tokenize(.....)
```
[the remark this answers is not in the document]
Comment (AlinaTR52) [reply lab_1_classify_by_unigrams/start.py]: done

**Pull request closed without merging**